Code/Framework/AzCore/AzCore/Component/ComponentApplication.cpp

```diff
@@ -1219,129 +1221,81 @@ namespace AZ
+        auto GemModuleVisitor = [&settingsRegistry = *m_settingsRegistry, &modulesLoadData]
+        (AZStd::string_view gemRootObjectPath, AZStd::string_view, Type)
+        {
+            auto VisitGemObjectFields = [&settingsRegistry, &modulesLoadData](AZStd::string_view jsonPath,
+                AZStd::string_view fieldName, Type)
             {
-                if (action == AZ::SettingsRegistryInterface::VisitAction::Begin)
+                AZStd::string_view gemModuleName = fieldName;
+                auto FindGemModuleLoadEntry = [gemModuleName](const GemModuleLoadData& moduleLoadData)
                 {
-                    // Strip off the last JSON pointer key from the path and if it matches the gem module key then add an entry
-                    // to the ModulesLoadData array
-                    AZStd::optional<AZStd::string_view> gemNameKey = AZ::StringFunc::TokenizeLast(path, "/");
-                    if (path == m_gemModuleKey && gemNameKey && !gemNameKey->empty())
-                    {
-                        AZStd::string_view gemName = gemNameKey.value();
-                        auto FindGemModuleLoadEntry = [gemName](const GemModuleLoadData& moduleLoadData)
-                        {
-                            return gemName == moduleLoadData.m_gemName;
-                        };
-
-                        if (auto foundIt = AZStd::find_if(m_modulesLoadData.begin(), m_modulesLoadData.end(), FindGemModuleLoadEntry);
-                            foundIt == m_modulesLoadData.end())
-                        {
-                            m_modulesLoadData.emplace_back(GemModuleLoadData{ gemName });
-                        }
-                    }
-                }
+                    return gemModuleName == moduleLoadData.m_gemModuleName;
+                };
 
-                return AZ::SettingsRegistryInterface::VisitResponse::Continue;
-            }
+                auto foundIt = AZStd::ranges::find_if(modulesLoadData, FindGemModuleLoadEntry);
+                GemModuleLoadData& moduleLoadData = foundIt != modulesLoadData.end()
+                    ? *foundIt
+                    : modulesLoadData.emplace_back(GemModuleLoadData{ gemModuleName });
 
-            using SettingsRegistryInterface::Visitor::Visit;
-            void Visit(AZStd::string_view path, AZStd::string_view valueName, AZ::SettingsRegistryInterface::Type, bool value) override
-            {
                 // By default the auto load option is true
-                // So auto load is turned off if option "AutoLoad" key is bool that is false
-                if (valueName == "AutoLoad" && !value)
+                // So auto load is turned off if option "AutoLoad" key exist and has a value of false
+                auto autoLoadJsonPath = FixedValueString::format("%.*s/AutoLoad",
+                    AZ_STRING_ARG(jsonPath));
+                if (bool autoLoadModule{}; settingsRegistry.Get(autoLoadModule, autoLoadJsonPath) && !autoLoadModule)
```

Comment:
Developers need to learn about [zero initialization](https://en.cppreference.com/w/cpp/language/zero_initialization).

Knowing about it, makes it a lot easier to just substitute the type of bool with `int` or any other type and the variable is still zero initialized.
> The effects of zero initialization are:

    If T is a scalar type, the object's initial value is the integral constant zero [explicitly converted](https://en.cppreference.com/w/cpp/language/explicit_cast) to T. 

    If T is an non-union class type, all base classes and non-static data members are zero-initialized, and all padding is initialized to zero bits. The constructors, if any, are ignored. 

    If T is a union type, the first non-static named data member is zero-initialized and all padding is initialized to zero bits. 

    If T is array type, each element is zero-initialized. 

    If T is reference type, nothing is done.

Reply:
I disagree with using zero initialization. 
`var x = 0` is basically the same in all languages and so makes O3DE more welcoming to developers coming to it where c++ is not their primary language. A full and complete understanding of c++20 should not be required to work on O3DE.

Reply:
Is zero initialization an optimization? Does var = 0 perform meaningfully worse?

Is there any reason to write code that is less obvious in what it does?

Reply:
> Is zero initialization an optimization?
Performance wise, no
Maintenance wise, yes
Let's say the
```
bool autoLoadModule{}
```
Later on the the autoLoadModule setting is changed to an string
```
AZStd::string autoLoadModule{}
```

The initialization of the variable doesn't have to change only the type

> Does var = 0 perform meaningfully worse?
The variable here isn't an int, it's bool

So that would be `var =false`.
But no it doesn't perform meaningfully worse

> Is there any reason to write code that is less obvious in what it does?
You mean like this [code](https://github.com/o3de/o3de/pull/8000/files#diff-20b18ccbf179ab34e487c8c8ad250aa3471429302707a13444a692b73211bc25R79)
What is the type of `auto` here?

Reply:
is 
`auto memoryView = m_memoryAllocator.Allocate(memoryRequirements.size, memoryRequirements.alignment);` 
less readable than
`AZ::Vulkan::MemoryView memoryView = m_memoryAllocator.Allocate(memoryRequirements.size, memoryRequirements.alignment);`
I would argue that the simpler version tells everyone what they need to know and imposes a minimum cognitive burden to understand.  

In your example, autoLoadModule's value comes from an api query, a change to a string would likely involve much larger changes. Changing to string initialization would be a minor concern for that level of change. 

My argument here is that `bool autoLoadModule = false;` is the most basic variable assignment all programmers learn. You're already doing that inside an 'if()' and using the result of a function call in the if inside the if test, that line is hard enough to follow on it's own, lets not add extra cognitive burden on the programmers that will be using O3DE. We're not here to teach them c++, we're here to provide an engine they can work with.

Reply:
This code isn't enforcing that other programmers using O3DE has to follow this approach.
I am fine with other contributors using whatever C++ construct they want to use to write correct code.

If another contributor wrote the following code, I am perfectly fine with it.
```c++
bool autoLoadModule = false;
if (settingsRegistry.Get(autoLoadModule, autoLoadJsonPath) && !autoLoadModule)
```
to initialize the value the of autoLoadModule(though this has a meaningful effect that the variable is on the stack for a larger scope).

What I am not find with is restricting a perfectly valid C++ construct, that has been around since standard C++ started, because of a reviewer not knowing about it,
I would NOT tell a contributor to do the following:
Theoretical Example: "Contributor can be please use use `bool autoLoadModule = bool{}` because that is how I would write it as a personal preference" 

Curly brace initialization isn't an esoteric feature that is barely used.

```c++
AZStd::vector<int> foo{};
// ... 
// clear vector and memory associated with it
foo = AZStd::vector<int>{};
```
Both those constructs are the same, the only difference is there being a variable name of "foo" there.

Reply:
It's not directly talking about this but I believe this section from the coding standard is relevant here

> Recommended: Make your names verbose. That way people will not need to guess, the code will be more readable and you will need less documentation. Typing is not your enemy, copy and paste is. Don't try to shorten words too much either. i, inst and instance are not the same and while inst could be common enough for people to know you mean instance, this is not always the case. Remember that programmers from many different domains might come across your code, try not to take anything for granted.

We should make our code easy for non-expert programmers to read.

For me the use of the `var{}` also fails on searchability, you need to already know about auto zero initialization to know the correct term to google.

Reply:
This is not really something that is expert programming knowledge.
The searchability argument doesn't hold: If someone saw `var{}`, they would most likely search for "curly brace initialization" or "[brace initialization](https://www.google.com/search?q=brace+initialization)", where on the first [result page](https://en.cppreference.com/w/cpp/language/aggregate_initialization) has a link to zero initialization within it.

Do you expect only expert programmers to know about initializer list?
Such as the following initialization of a vector
```
AZStd::vector<int>{5, 7};
AZStd::vector<int>{5};
AZStd::vector<int>{};
```

If a programmer has done the following, they have just used zero initialization
```
struct Foo
{
   int m_var;
};
Foo:::Foo()
   : m_var()
{
}
```

Reply:
Hey there, sorry I didn't mean for this to become a big point of discussion, I was just mentioning something that caught my eye (it was just an uber-nit I was happy for you to ignore).

I was coming at it from the same perspective @rgba16f was, that regular assignment is just much more common throughout the code base. I do get how `{}` can be super useful in a lot of cases for zero initializing structs without having to use a `memset` etc... And @lumberyard-employee-dm you're totally right it's completely valid to use here, this was more of just a style question thinking about O3DE in general.

I'm fine with leaving it for now and it's good people do know the language (for what it's worth I did know what that was doing 😅 I was just mentioning it from a convention point of view). Thanks for the discussion and if you want to catch-up about it feel free to ping me on Discord/Slack 👍